### PR TITLE
Composite refs

### LINKED ISSUE #239: Add support for composite terms as reference operands in the language
It would be nice if composite terms could be used as reference operands.

Today we don't support this in the parser:

```
data.example.p[[x,y]]
```

Authors have to write:

```
v = [x,y], data.example.p[v]
```

### PR DESCRIPTION
**Pull request by mmussomele**
Fixes #239

### CODE REVIEW

**Comment by mmussomele** [comment]
Fixes #239

File: ast/check.go [hunk 1]
```diff
@@ -534,6 +538,19 @@ func (rc *refChecker) checkRefLeaf(tpe types.Type, ref Ref, idx int) *Error {
...
+		if tpe == nil {
+			return nil
+		}
+
+		exist, ok := tpe.(*types.Set)
+		if !ok {
+			return nil
+		}
+
+		unify1(rc.env, SetTerm(head), exist, false)
```
Comment (mmussomele) [inline ast/check.go]: ah yea that would be much better

File: ast/check.go [hunk 2]
```diff
@@ -487,6 +487,9 @@ func (rc *refChecker) checkRef(curr *TypeEnv, node *typeTreeNode, ref Ref, idx i
 			return newRefErrInvalid(ref[0].Location, ref, idx, exist, types.S, getOneOfForNode(node))
 		}
 
+	case Array, Object, *Set:
+		rc.env.tree.PutOne(value, rc.env.Get(value))
```
Comment (mmussomele) [inline ast/check.go]: Yea I think you're right, this case just shouldn't exist. I'll just let it hit default

**Comment by mmussomele** [comment]
All the issues should be fixed now.

**Comment by tsandall** [comment]
This is close, but I think more testing is needed. Specifically, anywhere that was branching based on ref operand type needs to be reviewed. E.g., what happens if you write...

```
p[[x,y]] { ... }
```
And then query `p[[foo.bar.baz[_], "qux"]]`?

I imagine we need to have special handling for nested refs?

What about comprehensions nested inside the composites?

**Comment by mmussomele** [comment]
Ready for another look, fixed up up some checking issues and corrected eval to handle refs nested within composite operands. Comprehensions also work inside the composites.

File: ast/check.go [hunk 3]
```diff
@@ -534,6 +535,13 @@ func (rc *refChecker) checkRefLeaf(tpe types.Type, ref Ref, idx int) *Error {
 			}
 		}
 
+	case Array, Object, *Set:
+		// Composite references operands may only be used with a set.
+		if !unifies(tpe, types.NewSet(types.A)) {
+			return newRefErrInvalid(ref[0].Location, ref, idx, tpe, types.NewSet(types.A), nil)
+		}
+		unify1(rc.env, head, keys, false)
```
Comment (tsandall) [inline ast/check.go]: This should check if the head unified w/ the keys.

```
if !unify1(...) { // type error }
```

File: ast/check_test.go [hunk 4]
```diff
@@ -738,6 +743,14 @@ func TestCheckRefErrInvalid(t *testing.T) {
 			want:  types.S,
 			oneOf: []Value{String("p"), String("q")},
 		},
+		{
+			note:  "composite ref into non-set",
```
Comment (tsandall) [inline ast/check_test.go]: What about type errors with terms embedded within composite operands? This is related to the comment from above.

For example...

```
p = {[1],[2],[3]}
```

And then...`p[["foo"]]` or `p[{"a": b"}]`.

**Comment by mmussomele** [comment]
Comments addressed.

File: topdown/topdown.go [hunk 5]
```diff
@@ -1823,14 +1831,30 @@ func evalRefRuleResultRecSet(t *Topdown, set *ast.Set, ref, path ast.Ref, iter f
...
 			path = append(path, e)
+
 			if err := evalRefRuleResultRec(t, e.Value, tail, path, iter); err != nil {
 				return err
 			}
 			path = path[:len(path)-1]
 			t.Unbind(undo)
 		}
 		return nil
+	case ast.Array, ast.Object, *ast.Set:
```
Comment (tsandall) [inline topdown/topdown.go]: This entire function can be simplified to unify the set elements with the next element in the ref. Something like...

```
	head, tail := ref[0], ref[1:]
	k := head.Value
	for _, e := range *set {
		undo, err := evalEqUnify(t, e.Value, k, nil, func(t *Topdown) error {
			return evalRefRuleResultRec(t, PlugValue(k, t.Binding), tail, path.Append(e), iter)
		})
		if err != nil {
			return err
		}
		t.Unbind(undo)
	}
	return nil
```

Comment (mmussomele) [inline topdown/topdown.go]: 0.o it's like magic

Comment (mmussomele) [inline topdown/topdown.go]: done

**Comment by mmussomele** [comment]
Done

**Pull request merged**